test/gtest/ucm/malloc_hook.cc

```diff
@@ -0,0 +1,208 @@
+
+    /* Install memory hooks */
+    result = ucm_set_event_handler(UCM_EVENT_VM_MAPPED|UCM_EVENT_VM_UNMAPPED,
+                                   0, mem_event_callback,
+                                   reinterpret_cast<void*>(this));
+    ASSERT_UCS_OK(result);
+
+    /* Allocate small pointers with new heap manager */
+    for (int i = 0; i < small_alloc_count; ++i) {
+        new_ptrs.push_back(malloc(small_alloc_size));
+    }
+    EXPECT_GT(m_map_size, 0lu) << "thread " << m_thread_ind;
+    small_map_size = m_map_size;
+
+    num_ptrs_in_range = 0;
+    for (ucs::ptr_vector<void>::const_iterator iter = new_ptrs.begin();
+                    iter != new_ptrs.end(); ++iter)
+    {
+        if (is_ptr_in_range(*iter, small_alloc_size, m_map_ranges)) {
+            ++num_ptrs_in_range;
+        }
+    }
+    /* Need at least one ptr in the mapped ranges */
+    EXPECT_GT(num_ptrs_in_range, 0) << "thread " << m_thread_ind;
+
+    /* Allocate large chunk */
+    void *ptr = malloc(large_alloc_size);
+    EXPECT_GE(m_map_size, large_alloc_size + small_map_size) << "thread " << m_thread_ind;
+    EXPECT_TRUE(is_ptr_in_range(ptr, large_alloc_size, m_map_ranges)) << "thread " << m_thread_ind;
+
+    free(ptr);
+    EXPECT_GE(m_unmap_size, large_alloc_size) << "thread " << m_thread_ind;
+    /* coverity[pass_freed_arg] */
+    EXPECT_TRUE(is_ptr_in_range(ptr, large_alloc_size, m_unmap_ranges)) << "thread " << m_thread_ind;
+
+    void *s = strdup("test");
+    free(s);
+
+    /* Release old pointers (should not crash) */
+    old_ptrs.clear();
+
+    m_map_ranges.clear();
+    m_unmap_ranges.clear();
+
+    /* Don't release pointers before other threads exit, so they will map new memory
+     * and not reuse memory from other threads.
+     */
+    pthread_barrier_wait(m_barrier);
+
+    /* Release new pointers  */
+    new_ptrs.clear();
+
+    /* Call several malloc routines */
+    malloc_trim(0);
+
+    ptr = malloc(large_alloc_size);
+    if (!RUNNING_ON_VALGRIND) {
+        void *state = malloc_get_state();
+        malloc_set_state(state);
+    }
+    free(ptr);
+
+    pthread_mutex_lock(&lock);
+    UCS_TEST_MESSAGE << "thread " << m_thread_ind << "/" << m_num_threads
+                     << ": small mapped: " << small_map_size
+                     <<  ", total mapped: " << m_map_size
+                     <<  ", total unmapped: " << m_unmap_size;
+    std::cout.flush();
+    pthread_mutex_unlock(&lock);
+
+    ucm_unset_event_handler(UCM_EVENT_VM_MAPPED|UCM_EVENT_VM_UNMAPPED,
+                            mem_event_callback,
+                            reinterpret_cast<void*>(this));
+}
+
+UCS_TEST_F(malloc_hook, hook) {
+    static const int num_threads = 10;
+    ucs::ptr_vector<test_thread> threads;
+    pthread_barrier_t barrier;
+
+    pthread_barrier_init(&barrier, NULL, num_threads);
+    for (int i = 0; i < num_threads; ++i) {
+        threads.push_back(new test_thread(i, num_threads, &barrier));
+    }
+
+
+    threads.clear();
```

Comment:
should work because there is no state in a driver/hw. But i can add a test for it.